README.md
# triton-audit-logger

This package is an adaptation of the [restify audit logging
plugin](https://github.com/restify/node-restify/blob/master/lib/plugins/audit.js)
for use in Triton APIs. It adds a few features to the restify audit logger.
See the [top comment in lib/audit-logger.js](./lib/audit-logger.js) for
details. (This started with the restify v4 audit logger, so it is possible that
current restify versions do some of the same things.)

(This repository is part of the Joyent Triton project. See the [contribution
guidelines](https://github.com/joyent/triton/blob/master/CONTRIBUTING.md) --
*Triton does not use GitHub PRs* -- and general documentation at the main
[Triton project](https://github.com/joyent/triton) page.)


## Usage

Reasonably minimal usage (see [examples/hello-world.js](./examples/hello-world.js)):

```javascript
var tritonAuditLogger = require('triton-audit-logger');
var bunyan = require('bunyan');
var restify = require('restify');

var NAME = 'hello-world';
var log = bunyan.createLogger({
    name: NAME,
    serializers: restify.bunyan.serializers
});
var server = restify.createServer({
    name: NAME,
    log: log,
    // ...
});

server.on('after', tritonAuditLogger.createAuditLogHandler({
    log: log,
    // ...
}));
```

Suggested starter usage for Triton APIs:

```javascript
server.on('after', tritonAuditLogger.createAuditLogHandler({
    log: log,
    // Enable logging of request and response bodies. By default these are
    // clipped at 10k and response bodies for HTTP status 2xx are skipped
    // (because they are typically large and less interesting).
    reqBody: {},
    resBody: {},

    // Possibly include some overrides for particular routes. E.g.:
    routeOverrides: {
        // Never log 'ping' route requests.
        'getping': {include: false},

        // Reduce logging for a possibly frequent restify route at a
        // different level that isn't typically enabled.
        'getconfigs': {
            logLevel: 'debug'
        }
    }
}));
```

An example showing usage of most of the configuration options
(see the `createAuditLogHandler` comment for option docs):


```javascript
server.on('after', tritonAuditLogger.createAuditLogHandler({
    log: log,
    reqBody: {
        include: true,
        maxLen: 1024
    },
    resBody: {
        // Log the response body, but not for 200-299 HTTP status response.
        include: true,
        includeGet2xx: false
    },
    routeOverrides: {
        // Never log 'ping' route requests.
        'getping': {include: false},

        // Reduce logging for a possibly frequent restify route at a
        // different level that isn't typically enabled.
        'getconfigs': {
            logLevel: 'debug'
        }
    },
    polish: function censorAuth(fields, req, res, route, err) {
        // Censor potential auth info in a particular header.
        if (req.headers['x-registry-auth'] !== undefined) {
            req.headers['x-registry-auth'] = '(censored)';
        }
    }
}));
```


## Development

### Commiting

Before commit, ensure that the following passes:

    make fmt check

You can setup a local git pre-commit hook that'll do that by running

    make git-hooks

Also see the note at the top that https://cr.joyent.us is used for code review
for this repo.


### Releasing

Changes with possible user impact should:

1. Add a note to the [changelog](./CHANGES.md).
2. Bump the package version appropriately (major for breaking changes, minor
   for new features, patch for bug fixes).
3. Once merged to master, the new version should be tagged and published to npm
   via:

        make cutarelease

   To list to npm accounts that have publish access:

        npm owner ls $packageName


CHANGES.md
# triton-audit-logger Changelog

## not yet released

(nothing yet

## 1.0.1

- issue #1: Fix an errant logged "null" for `res.body` with restify v4.

## 1.0.0

First release.


test/unit/testcommon.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright (c) 2018, Joyent, Inc.
 */

'use strict';

const assert = require('assert-plus');
const bunyan = require('bunyan');
const restify = require('restify');
const restifyErrors = require('restify-errors');

const auditLogger = require('../..');

function deepCopy(obj) {
    return JSON.parse(JSON.stringify(obj));
}

function CapturingStream(recs) {
    this.recs = recs;
}
CapturingStream.prototype.write = function write(rec) {
    this.recs.push(rec);
};

/*
 * Creates a test restify server with the given audit logger options.
 *
 * It captures its bunyan log records. Use the following to access and clear
 * the recs. `getRecs` is async to allow the restify 'after' event to fire.
 * Otherwise, after a client response, the audit logger might not yet have
 * logged.
 *
 *      server.getRecs((recs) => { ... });
 *      server.clearRecs();
 *
 * The server has a few default endpoints:
 *
 * - `GET /hello`, JSON object response, 200 status.
 * - `PUT /join`, expects JSON body with `login` field, 200 or 409 status.
 * - `GET /oops`, 500 status, JSON error body. Should log the error.
 */
function createServer(opts) {
    assert.string(opts.name, 'opts.name');
    assert.optionalObject(opts.auditLoggerOpts, 'opts.auditLoggerOpts');

    let auditLoggerOpts = opts.auditLoggerOpts
        ? deepCopy(opts.auditLoggerOpts)
        : {};
    let log;
    let recs = [];
    let server;

    log = bunyan.createLogger({
        name: opts.name,
        serializers: restify.bunyan.serializers,
        streams: [
            {
                stream: new CapturingStream(recs),
                type: 'raw'
            }
        ]
    });

    server = restify.createServer({
        name: opts.name,
        log: log
    });

    server.getRecs = function getRecs(cb) {
        // Async to allow 'after' event to be handled for audit logging.
        setImmediate(() => {
            cb(recs);
        });
    };
    server.clearRecs = function clearRecs() {
        recs.length = 0;
    };

    auditLoggerOpts.log = log;
    server.on('after', auditLogger.createAuditLogHandler(auditLoggerOpts));

    // Endpoints.
    server.get('/hello', function hello(req, res, next) {
        res.send({hello: 'world'});
        next();
    });
    server.put(
        '/join',
        restify.plugins.bodyParser({mapParams: false}),
        function join(req, res, next) {
            if (!req.body.login) {
                next(
                    new restifyErrors.InvalidArgumentError(
                        'missing login field'
                    )
                );
            } else {
                res.send({success: true, login: req.body.login});
                next();
            }
        }
    );
    server.get({path: '/oops', name: 'oops'}, function oops(req, res, next) {
        let err = new Error('this was the root cause');
        next(new restifyErrors.InternalError(err, 'something blew up'));
    });

    return server;
}

module.exports = {
    createServer: createServer
};


test/unit/hello-world.test.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright (c) 2018, Joyent, Inc.
 */

'use strict'

const assert = require('assert-plus');
const bunyan = require('bunyan');
const restify = require('restify');
const restifyClients = require('restify-clients');
const restifyErrors = require('restify-errors');
const tap = require('tap');

const auditLogger = require('../..');

// ---- data

const addr = '127.0.0.1';
const port = '8123';
const url = 'http://' + addr + ':' + port;

// ---- support stuff

function deepCopy(obj) {
    return JSON.parse(JSON.stringify(obj));
}

/*
 * Creates a test restify server with the given audit logger options.
 *
 * It captures its bunyan log records. Use the following to access and clear
 * the recs. `getRecs` is async to allow the restify 'after' event to fire.
 * Otherwise, after a client response, the audit logger might not yet have
 * logged.
 *
 *      server.getRecs((recs) => { ... });
 *      server.clearRecs();
 *
 * The server has a few default endpoints:
 *
 * - `GET /hello`, JSON object response, 200 status.
 * - `PUT /join`, expects JSON body with `login` field, 200 or 409 status.
 * - `GET /oops`, 500 status, JSON error body. Should log the error.
 */
function createServer(opts) {
    assert.string(opts.name, 'opts.name');
    assert.optionalObject(opts.auditLoggerOpts, 'opts.auditLoggerOpts');

    let auditLoggerOpts = opts.auditLoggerOpts ?
            deepCopy(opts.auditLoggerOpts) : {};
    let log;
    let recs = [];
    let server;

    log = bunyan.createLogger({
        name: opts.name,
        serializers: restify.bunyan.serializers,
        streams: [
            {
                stream: new CapturingStream(recs),
                type: 'raw'
            }
        ]
    });

    server = restify.createServer({
        name: opts.name,
        log: log
    });

    server.getRecs = function getRecs(cb) {
        // Async to allow 'after' event to be handled for audit logging.
        setImmediate(() => {
            cb(recs);
        });
    }
    server.clearRecs = function clearRecs() {
        recs.length = 0;
    }

    auditLoggerOpts.log = log;
    server.on('after', auditLogger.createAuditLogHandler(auditLoggerOpts));

    // Endpoints.
    server.get('/hello', function hello(req, res, next) {
        res.send({'hello': 'world'});
        next();
    });
    server.put('/join',
        restify.plugins.bodyParser({mapParams: false}),
        function join(req, res, next) {
            if (!req.body.login) {
                next(new restifyErrors.InvalidArgumentError(
                    'missing login field'));
            } else {
                res.send({'success': true, 'login': req.body.login});
                next();
            }
        }
    );
    server.get({path: '/oops', name: 'oops'}, function oops(req, res, next) {
        let err = new Error('this was the root cause');
        next(new restifyErrors.InternalError(err, 'something blew up'));
    });

    return server;
}

function CapturingStream(recs) {
    this.recs = recs;
}
CapturingStream.prototype.write = function (rec) {
    this.recs.push(rec);
}

// ---- tests

tap.test('empty-opts', (tt) => {
    // Default audit logger config will exclude req/res bodies.
    const auditLoggerOpts = null;

    const client = restifyClients.createJsonClient({url: url});
    const server = createServer({
        name: 'empty-opts',
        auditLoggerOpts: auditLoggerOpts
    });

    tt.test('server listen', (t) => {
        server.listen(port, addr, function listening() {
            t.end();
        });
    });

    tt.test('GET /hello', (t) => {
        server.clearRecs();
        client.get('/hello', function (err, req, res, body) {
            t.ifErr(err, err);
            t.equal(res.statusCode, 200, '200 status code');
            server.getRecs((recs) => {
                var rec = recs[0]; // assuming audit rec is the only one
                t.ok(rec.audit, 'rec.audit');
                t.equal(rec.res.statusCode, 200, 'rec.res.statusCode');
                t.notOk(rec.res.body, 'no rec.res.body');
                t.end();
            });
        });
    });

    tt.test('PUT /join', (t) => {
        server.clearRecs();
        client.put('/join', {login: 'bob'}, function (err, req, res, body) {
            t.ifErr(err, err);
            t.equal(res.statusCode, 200, '200 status code');
            server.getRecs((recs) => {
                var rec = recs[0]; // assuming audit rec is the only one
                t.ok(rec.audit, 'rec.audit');
                t.notOk(rec.res.body, 'no rec.res.body');
                t.end();
            });
        });
    });

    tt.test('404', (t) => {
        server.clearRecs();
        client.get('/no-such-endpoint', function (err, req, res, body) {
            t.ok(err, 'err: ' + err);
            t.equal(res.statusCode, 404, '404 status code');
            server.getRecs((recs) => {
                var rec = recs[0]; // assuming audit rec is the only one
                t.ok(rec.audit, 'rec.audit');
                t.equal(rec.res.statusCode, 404, 'rec.res.statusCode');
                t.equal(rec.err.name, 'ResourceNotFoundError');
                t.notOk(rec.res.body, 'no rec.res.body');
                t.end();
            });
        });
    });

    tt.test('GET /oops', (t) => {
        server.clearRecs();
        client.get('/oops', function (err, req, res, body) {
            t.ok(err, 'err: ' + err);
            t.equal(res.statusCode, 500, '500 status code');
            server.getRecs((recs) => {
                var rec = recs[0]; // assuming audit rec is the only one
                t.ok(rec.audit, 'rec.audit');
                t.equal(rec.res.statusCode, 500, 'rec.res.statusCode');
                t.notOk(rec.res.body, 'no rec.res.body');
                t.equal(rec.err.name, 'InternalError');
                t.equal(rec.err.message, 'something blew up');
                t.ok(/root cause/.test(rec.err.stack),
                    'err stack includes "root cause" error');
                t.end();
            });
        });
    });

    tt.test('teardown', (t) => {
        server.close();
        client.close();
        t.end();
    });

    tt.end();
});


tap.test('default-body-logging', (tt) => {
    // By default resBody/reqBody = {} will: log response bodies, excluding
    // buffers, excluding responses with GET requests with a 2xx status code,
    // and will clip at 10k characters.
    const auditLoggerOpts = {
        resBody: {},
        reqBody: {}
    };

    const client = restifyClients.createJsonClient({url: url});
    const server = createServer({
        name: 'default-body-logging',
        auditLoggerOpts: auditLoggerOpts
    });

    tt.test('server listen', (t) => {
        server.listen(port, addr, function listening() {
            t.end();
        });
    });

    tt.test('GET /hello', (t) => {
        server.clearRecs();
        client.get('/hello', function (err, req, res, body) {
            server.getRecs((recs) => {
                var rec = recs[0]; // assuming audit rec is the only one
                t.ok(rec.audit, 'rec.audit');
                t.equal(rec.res.statusCode, 200, 'rec.res.statusCode');
                t.notOk(rec.res.body, 'no rec.res.body');
                t.end();
            });
        });
    });

    tt.test('PUT /join', (t) => {
        server.clearRecs();
        client.put('/join', {login: 'bob'}, function (err, req, res, body) {
            server.getRecs((recs) => {
                var rec = recs[0]; // assuming audit rec is the only one
                t.ok(rec.audit, 'rec.audit');
                t.equal(rec.res.statusCode, 200, 'rec.res.statusCode');
                t.equal(rec.req.body, '{"login":"bob"}');
                t.equal(rec.res.body, '{"success":true,"login":"bob"}');
                t.end();
            });
        });
    });

    tt.test('404', (t) => {
        server.clearRecs();
        client.get('/no-such-endpoint', function (err, req, res, body) {
            t.equal(res.statusCode, 404, '404 status code');
            server.getRecs((recs) => {
                var rec = recs[0]; // assuming audit rec is the only one
                t.ok(rec.audit, 'rec.audit');
                t.equal(rec.res.statusCode, 404, 'rec.res.statusCode');
                t.ok(rec.res.body, 'rec.res.body');
                t.end();
            });
        });
    });

    tt.test('GET /oops', (t) => {
        server.clearRecs();
        client.get('/oops', function (err, req, res, body) {
            t.equal(res.statusCode, 500, '500 status code');
            server.getRecs((recs) => {
                var rec = recs[0]; // assuming audit rec is the only one
                t.ok(rec.audit, 'rec.audit');
                t.equal(rec.res.statusCode, 500, 'rec.res.statusCode');
                t.ok(rec.res.body, 'rec.res.body');
                t.ok(/Internal/.test(rec.res.body),
                    '"Internal" in rec.res.body');
                t.ok(/something blew up/.test(rec.res.body),
                    '"something blew up" in rec.res.body');
                t.end();
            });
        });
    });

    tt.test('teardown', (t) => {
        server.close();
        client.close();
        t.end();
    });

    tt.end();
});


examples/hello-world.js
//
// Start the server and note the logging output for requests:
//      node examples/hello-world.js | bunyan
//
// Then in a separate terminal make a few requests against it:
//      curl -i http://127.0.0.1:8000/bogus    # expect a 404
//      curl -i http://127.0.0.1:8000/hello    # expect a 200
//      curl -i http://127.0.0.1:8000/oops     # expect a 500, log at ERROR lvl
//

var tritonAuditLogger = require('../');
var bunyan = require('bunyan');
var errors = require('restify-errors');
var restify = require('restify');

var NAME = 'hello-world';
var log = bunyan.createLogger({
    name: NAME,
    serializers: restify.bunyan.serializers
});
var server = restify.createServer({
    name: NAME,
    log: log
    // ...
});

server.on(
    'after',
    tritonAuditLogger.createAuditLogHandler({
        log: log,
        resBody: {},
        routeOverrides: {
            oops: {
                logLevel: 'error'
            }
        }
    })
);

// Add some endpoints for play.
server.get('/hello', function hi(req, res, next) {
    res.send({hello: 'world'});
    next();
});
server.get({path: '/oops', name: 'oops'}, function hi(req, res, next) {
    var err = new Error('boom');
    next(new errors.InternalError(err, 'something blew up'));
});

var PORT = process.env.PORT || 8000;
var ADDR = '127.0.0.1';
server.listen(PORT, ADDR, function listening() {
    log.info('listening (http://%s:%s)', ADDR, PORT);
});


lib/audit-logger.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright (c) 2018, Joyent, Inc.
 */

/*
 * This is an adaptation of restify/lib/plugins/audit.js (originally from
 * restify 4.x) with a number of feature additions.
 *
 * Usage:
 *
 *      server.on('after', audit.createAuditLogHandler({
 *          log: log,
 *          ...options...
 *      }));
 *
 * Features and differences:
 *
 * - Fine control on how and whether to log request and response bodies.
 *   `opts.reqBody` and `opts.resBody` can be used to independently control
 *   how the request and response bodies are logged. At its simplest,
 *   `{resBody: {}}`, will log response bodies, excluding buffers, excluding
 *   responses with GET requests with a 2xx status code, and will clip
 *   at 10k characters.
 *
 *   These replace `opts.body` from the default restify audit logger.
 *
 * - This audit logger attempts to log the req/res bodies as close to the wire
 *   as possible. `res.body` is the *formatted* response body, as opposed
 *   to the default restify audit logger which logs the response body *object*
 *   before formatting. `req.body` is the raw request body *only for restify
 *   5.x and later* (when restify made `req.rawBody` available).
 *
 * - Logged records set `audit: true`.
 *   (Note that restify's audit logger sets that *and* `_audit: true`.)
 *
 * - There is an optional `polish` function that can be passed in to customize
 *   audit log records just before they are logged by Bunyan.
 *
 * - Most options can be overridden per route name via `opts.routeOverrides`.
 *
 * See the `createAuditLogHandler` block comment for specifics.
 */

var assert = require('assert-plus');

// Default maximum length for requests/responses body that are logged.
var DEFAULT_BODY_MAX_LEN = 10 * 1024;

// Avoid a bunyan package dependency by copying some code from Bunyan 1.8.12.
var bunyanErrSerializer;
var bunyanNameFromLevel;
var bunyanResolveLevel;
try {
    var bunyan = require('bunyan');
    bunyanErrSerializer = bunyan.stdSerializers.err;
    bunyanNameFromLevel = bunyan.nameFromLevel;
    bunyanResolveLevel = bunyan.resolveLevel;
} catch (_bunyanImportErr) {
    var format = require('util').format;

    var getFullErrorStack = function _getFullErrorStack(ex) {
        var ret = ex.stack || ex.toString();
        if (ex.cause && typeof ex.cause === 'function') {
            var cex = ex.cause();
            if (cex) {
                ret += '\nCaused by: ' + getFullErrorStack(cex);
            }
        }
        return ret;
    };

    bunyanErrSerializer = function _errSerializer(err) {
        if (!err || !err.stack) {
            return err;
        }
        var obj = {
            message: err.message,
            name: err.name,
            stack: getFullErrorStack(err),
            code: err.code,
            signal: err.signal
        };
        return obj;
    };

    var TRACE = 10;
    var DEBUG = 20;
    var INFO = 30;
    var WARN = 40;
    var ERROR = 50;
    var FATAL = 60;

    var levelFromName = {
        trace: TRACE,
        debug: DEBUG,
        info: INFO,
        warn: WARN,
        error: ERROR,
        fatal: FATAL
    };
    bunyanNameFromLevel = {};
    Object.keys(levelFromName).forEach(function(name) {
        bunyanNameFromLevel[levelFromName[name]] = name;
    });

    bunyanResolveLevel = function resolveLevel(nameOrNum) {
        var level;
        var type = typeof nameOrNum;
        if (type === 'string') {
            level = levelFromName[nameOrNum.toLowerCase()];
            if (!level) {
                throw new Error(format('unknown level name: "%s"', nameOrNum));
            }
        } else if (type !== 'number') {
            throw new TypeError(
                format(
                    'cannot resolve level: invalid arg (%s):',
                    type,
                    nameOrNum
                )
            );
        } else if (nameOrNum < 0 || Math.floor(nameOrNum) !== nameOrNum) {
            throw new TypeError(
                format('level is not a positive integer: %s', nameOrNum)
            );
        } else {
            level = nameOrNum;
        }
        return level;
    };
}

function auditBodyFromReq(req, bodyOpts) {
    var auditBody;

    if (bodyOpts === undefined) {
        return undefined;
    }

    // We want the audit log to be as close as possible to the what is on
    // the wire. Restify 5.x added support for `req.rawBody`
    // (https://github.com/restify/node-restify/issues/928). We'll use that
    // if available.
    auditBody = req.hasOwnProperty('rawBody') ? req.rawBody : req.body;

    if (!bodyOpts.include) {
        auditBody = undefined;
    } else if (!auditBody) {
        /* jsl:pass */
    } else if (Buffer.isBuffer(auditBody) && !bodyOpts.includeBuffers) {
        auditBody = '<buffer>';
    } else if (
        typeof auditBody === 'string' &&
        auditBody.length > bodyOpts.maxLen
    ) {
        auditBody =
            auditBody.slice(0, bodyOpts.maxLen) +
            '\n...<elided ' +
            (auditBody.length - bodyOpts.maxLen) +
            ' chars>';
    }

    return auditBody;
}

// Same as `auditBodyFromReq` plus:
// - `req._body` is trickier
// - handle `bodyOpts.includeGet2xx`.
function auditBodyFromRes(res, bodyOpts) {
    var auditBody;

    if (bodyOpts === undefined) {
        return undefined;
    }

    // We want the audit log to be as close as possible to the what is on
    // the wire. That means we prefer the *formatted* response body (i.e. after
    // the restify "formatter" has been applied). `res._data` holds this --
    // at least that appears to be so in restify 4.x and 6.x code.
    //
    // This differs from the core restify audit logger that uses the
    // *unformatted* response body:
    //    if (res._body instanceof HttpError) {
    //        auditBody = res._body.body;
    //    } else {
    //        auditBody = res._body;
    //    }
    //
    // One reason we want the formatted body is that we can more realistically
    // apply `maxLen`.
    auditBody = res._data;

    // In restify v4 (but not in restify v7), `res._data` would be set to
    // `null` rather than a more appropriate `undefined`. Handle that here.
    // Dev Note: If this proves problematic for an intended `null` body,
    // then lets remove this hack for older restify versions.
    if (auditBody === null) {
        auditBody = undefined;
    }

    if (!bodyOpts.include) {
        auditBody = undefined;
    } else if (
        !bodyOpts.includeGet2xx &&
        res.req.method === 'GET' &&
        (res.statusCode >= 200 && res.statusCode < 300)
    ) {
        auditBody = undefined;
    } else if (!auditBody) {
        /* jsl:pass */
    } else if (Buffer.isBuffer(auditBody) && !bodyOpts.includeBuffers) {
        auditBody = '<buffer>';
    } else if (
        typeof auditBody === 'string' &&
        auditBody.length > bodyOpts.maxLen
    ) {
        auditBody =
            auditBody.slice(0, bodyOpts.maxLen) +
            '\n...<elided ' +
            (auditBody.length - bodyOpts.maxLen) +
            ' chars>';
    }

    return auditBody;
}

/**
 * Manually generates a POJO from `res.getHeaderNames` and `res.getHeader`,
 * if available, falling back to deprecated `res._headers`, otherwise.
 * Intentionally does not use `res.getHeaders` to avoid deserialization
 * issues with object returned by that method.
 *
 * See https://github.com/restify/node-restify/issues/1370
 *
 * (This was lifted from restify v6.x's lib/plugins/audit.js.)
 */
function getResponseHeaders(res) {
    if (res.getHeaderNames && res.getHeader) {
        return res.getHeaderNames().reduce(function reduce(prev, curr) {
            var header = {};
            header[curr] = res.getHeader(curr);
            return Object.assign({}, prev, header);
        }, {});
    }
    return res._headers;
}

/*
 * A "bodyOpts" is one of the `opts.reqBody` or `opts.resBody` options to
 * `createAuditLogHandler`.
 */
function assertOptionalBodyOpts(bo, bodyName) {
    if (!bo) {
        return;
    }
    assert.optionalBool(bo.include, bodyName + '.include');
    assert.optionalBool(bo.includeBuffers, bodyName + '.includeBuffers');
    assert.optionalBool(bo.includeGet2xx, bodyName + '.includeGet2xx');
    assert.optionalNumber(bo.maxLen, bodyName + '.maxLen');
}

/*
 * A `routeOpts` is one of the objects specified as values to
 * `opts.routeOverrides`. The top-level `opts` to `createAuditLogHandler`
 * is also a superset of `routeOpts`.
 */
function assertRouteOpts(ro, routeName) {
    assert.optionalBool(ro.include, routeName + '.include');
    // logLevel is validated in `Normalize` section below.
    assertOptionalBodyOpts(ro.reqBody, routeName + '.reqBody');
    assertOptionalBodyOpts(ro.resBody, routeName + '.resBody');
    assert.optionalFunc(ro.polish, routeName + '.polish');
}

function normalizeRouteOpts(ro, defaults) {
    // - default 'ro.include' to true
    if (ro.include === undefined) {
        ro.include = true;
    }
    // - set `ro.logFnName`, default to "info"
    if (ro.logLevel) {
        ro.logFnName = bunyanNameFromLevel[bunyanResolveLevel(ro.logLevel)];
        assert.string(
            ro.logFnName,
            'can resolve logLevel="' +
                ro.logLevel +
                '" to a Bunyan log level name'
        );
    } else {
        ro.logFnName = 'info';
    }
    [ro.reqBody, ro.resBody].forEach(function(bodyOpt) {
        if (bodyOpt) {
            // - default `reqBody.include` and `resBody.include` to true
            if (!bodyOpt.hasOwnProperty('include')) {
                bodyOpt.include = true;
            }
            // - default `reqBody.maxLen` and `resBody.maxLen`
            if (!bodyOpt.hasOwnProperty('maxLen')) {
                bodyOpt.maxLen = DEFAULT_BODY_MAX_LEN;
            }
        }
    });
    // - inherit values from `defaultRouteOpts`
    if (defaults) {
        Object.keys(defaults).forEach(function(fieldName) {
            if (ro[fieldName] === undefined) {
                ro[fieldName] = defaults[fieldName];
            }
        });
    }
}

// ---- API

/**
 * Create an audit log handler.
 *
 * @param {Object} opts:
 *      @param {Object} opts.log - A Bunyan logger on which to log. Required.
 *      @param {Boolean} opts.include - Whether to log at all. Default true.
 *      @param {String} opts.logLevel - The bunyan log level (either the name
 *          or the integer value) at which to log. Default is "info".
 *      @param {Object} opts.reqBody - Options for logging request bodies.
 *      @param {Object} opts.resBody - Options for logging response bodies.
 *          Each of `reqBody` and `resBody` may have the following fields.
 *          All fields are optional.
 *          - {Boolean} `include` - Whether to log the body. Default is true.
 *          - {Number} `maxLen` - The maximum length of body to log.
 *            Default is 10k.
 *          - {Boolean} `includeBuffers` - Whether to log the body even if it
 *            is a Buffer. By default buffers are logged as `<buffer>`.
 *          - {Boolean} `includeGet2xx` - Whether to include response bodies
 *            for "GET" requests with a success response code (i.e. 2xx).
 *            Default is false. The reasoning is that successful GET response
 *            bodies can tend to be large and uninteresting. This field is
 *            only relevant for `opts.resBody`.
 *      @param {Function} opts.polish - A *sync* function that is called for
 *          each audit log record just before the call to log it. Called as:
 *              `function (fields, req, res, route, err)`
 *          where `fields` is the Bunyan fields object being logged. It
 *          can be changed in-place (as could `req` et al) to "polish" the
 *          audit log record.
 *      @param {Object} opts.routeOverrides - A mapping of `route.name`
 *          (recall that restify, at least v4.x, lowercases `route.name`), to
 *          overrides for any of the above options, except `log`. E.g.:
 *              {
 *                  // Exclude "GetPing" requests.
 *                  'getping': {include: false}
 *                  // Log "GetConfig" requests at DEBUG-level.
 *                  'getconfig': {logLevel: 'debug'}
 *                  // See the first 300 chars of ListInstances responses.
 *                  'listinstances': {
 *                      resBody: {
 *                          includeGet2xx: true,
 *                          maxLen: 300
 *                      }
 *                  }
 *              }
 *          Dev Note: This function modifies the `routeOverrides` objects
 *          in-place.
 * @returns {Function} A restify handler intended for `server.on('after', ...)`.
 */
function createAuditLogHandler(opts) {
    // Validate inputs.
    assert.object(opts, 'opts');
    assert.object(opts.log, 'opts.log');
    assert.optionalObject(opts.routeOverrides, 'opts.routeOverrides');
    var defaultRouteOpts = {
        include: opts.include,
        logLevel: opts.logLevel,
        reqBody: opts.reqBody,
        resBody: opts.resBody,
        polish: opts.polish
    };
    assertRouteOpts(defaultRouteOpts, 'default');
    var routeOptsFromName = opts.routeOverrides || {};
    Object.keys(routeOptsFromName).forEach(function(name) {
        assertRouteOpts(routeOptsFromName[name], 'routeOverrides.' + name);
    });

    // Normalize routeOpts objects.
    normalizeRouteOpts(defaultRouteOpts);
    Object.keys(routeOptsFromName).forEach(function(name) {
        normalizeRouteOpts(routeOptsFromName[name], defaultRouteOpts);
    });

    function routeOptsFromRoute(route) {
        return (
            (route && route.name && routeOptsFromName[route.name]) ||
            defaultRouteOpts
        );
    }

    var errSerializer =
        opts.log.serializers && opts.log.serializers.err
            ? opts.log.serializers.err
            : bunyanErrSerializer;
    var log = opts.log.child({
        audit: true,
        serializers: {
            err: errSerializer,
            req: function auditReqSerializer(req) {
                if (!req) {
                    return false;
                }

                var routeOpts = routeOptsFromRoute(req.route);

                var timers = {};
                (req.timers || []).forEach(function(time) {
                    var t = time.time;
                    var _t = Math.floor(1000000 * t[0] + t[1] / 1000);
                    // TODO: restify 6.x diff here to consider
                    timers[time.name] = _t;
                });

                return {
                    body: auditBodyFromReq(req, routeOpts.reqBody),
                    // TODO: consider connectionState from restify 6.x here
                    headers: req.headers,
                    httpVersion: req.httpVersion,
                    method: req.method,
                    // account for native and queryParser plugin usage
                    query:
                        typeof req.query === 'function'
                            ? req.query()
                            : req.query,
                    timers: timers,
                    trailers: req.trailers,
                    url: req.url,
                    version: req.version()
                };
            },
            res: function auditResSerializer(res) {
                if (!res) {
                    return false;
                }

                var routeOpts = routeOptsFromRoute(res.req.route);

                return {
                    body: auditBodyFromRes(res, routeOpts.resBody),
                    headers: getResponseHeaders(res),
                    statusCode: res.statusCode,
                    trailer: res._trailer || false
                };
            }
        }
    });

    function audit(req, res, route, err) {
        var routeName = (route && route.name) || undefined;
        var routeOpts = routeOptsFromRoute(route);

        if (!routeOpts.include) {
            return;
        }

        // TODO: 6.x diff here to accomodate.
        var latency = res.get('Response-Time');
        if (typeof latency !== 'number') {
            latency = Date.now() - req._time;
        }

        var fields = {
            err: err,
            latency: latency,
            remoteAddress: req.connection.remoteAddress,
            remotePort: req.connection.remotePort,
            req: req,
            req_id: req.getId(),
            res: res,
            route: routeName,
            secure: req.secure
        };

        if (routeOpts.polish) {
            routeOpts.polish(fields, req, res, route, err);
        }

        log[routeOpts.logFnName](fields, 'handled: %d', res.statusCode);
    }

    return audit;
}

// ---- exports

module.exports = {
    createAuditLogHandler: createAuditLogHandler
};
